File: Chapter 03/Python/section_3_4_order_processor_refactor/before/order_processor.py
from .order import Order

"""
ANTI-PATTERN: The "God Class" / Monolithic Script.
* ARCHITECTURE PROBLEM: This class is the "Everything Hub." It violates 
the Single Responsibility Principle (SRP) by managing four distinct 
responsibilities:
1. Validation Logic
2. Payment Gateway Integration
3. Inventory State Management
4. Notification / Communication Logic

WHY THIS IS BRITTLE:
- Hard to Test: You can't test if the validation logic works without 
  triggering print statements for payments and emails.
- High Coupling: If you switch from Email to SMS notifications, you have 
  to modify this core business processing file.
- Fragile: A failure in the "Notify" step could theoretically crash 
  the entire order process if not handled correctly.
"""

class OrderProcessor:
    def process(self, order: 'Order'):
        # 🚨 ARCHITECTURE WARNING: Responsibility 1 - Validation.
        # This logic should be encapsulated in a dedicated validator.
        print("  [Validate] Validating order...")
        if not order.items or order.total <= 0:
            raise ValueError("Order is invalid.")

        # 🚨 ARCHITECTURE WARNING: Responsibility 2 - Payment.
        # This couples the high-level flow to specific billing logic.
        print(f"  [Payment] Processing payment for ${order.total:.2f}...")
        payment_success = True

        if payment_success:
            # 🚨 ARCHITECTURE WARNING: Responsibility 3 - Inventory.
            # Direct dependency on infrastructure/database concerns.
            print("  [Inventory] Updating inventory...")

            # 🚨 ARCHITECTURE WARNING: Responsibility 4 - Notifications.
            # Direct dependency on external communication services.
            print(f"  [Notify] Sending confirmation email to {order.customer_email}...")
            
            return "Order processed successfully."
        else:
            return "Payment failed."

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_4_order_processor_refactor/before/OrderProcessor.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_4_order_processor_refactor.before;

/**
 * ANTI-PATTERN: The "God Class" / Monolithic Transaction Script.
 * * ARCHITECTURE PROBLEM: This class is a massive violation of the Single 
 * Responsibility Principle (SRP) and is tightly coupled to four distinct domains.
 * * If this were a real enterprise application, this one method would:
 * 1. Contain complex business validation rules.
 * 2. Talk to a third-party Payment Gateway API (like Stripe).
 * 3. Connect to a SQL Database to update inventory tables.
 * 4. Talk to a third-party Email API (like SendGrid).
 * * WHY THIS FAILS:
 * - Testing: You cannot test the validation logic without accidentally triggering 
 * real payment calls or database updates.
 * - Maintenance: Four different teams (Billing, Inventory, Communications, Sales) 
 * will constantly be modifying this exact same file, causing merge conflicts.
 * - Brittleness: A timeout error from the Email API could cause the entire 
 * payment process to roll back or crash!
 */
public class OrderProcessor {
    public String process(Order order) {
        
        // 🚨 ARCHITECTURE WARNING: Domain Rule Violation
        // 1. Validation
        System.out.println("  [Validate] Validating order...");
        if (order.items.isEmpty() || order.total <= 0) {
            throw new IllegalStateException("Order is invalid.");
        }

        // 🚨 ARCHITECTURE WARNING: External Integration Coupling
        // 2. Payment Processing
        System.out.println("  [Payment] Processing payment for $" + order.total + "...");
        boolean paymentSuccess = true;

        // 🚨 ARCHITECTURE WARNING: Infrastructure & Communications Coupling
        // 3. Inventory Update & 4. Confirmation Email
        if (paymentSuccess) {
            System.out.println("  [Inventory] Updating inventory...");
            System.out.println("  [Notify] Sending confirmation email to " + order.customerEmail + "...");
            return "Order processed successfully.";
        } else {
            return "Payment failed.";
        }
    }
}